websocket implementation

siddarthvader opened the pull request:
implemneted websockets networking layer along with HTTP for hooks invocation rather we have bidirectional communication now between compiler and orchestrator. There is no schema change, DB already has a column for port in plugin table.


 
**Since we have plugin to plugin hook invocation also, we are using our existing http layer for that, we have abstracted the hook handlers into a new files called handlers.go and calling these functions form both http and websocket layer.**

We maintain a map of websocket connection, and we keep an account of pending request, this is still request-response model nothing inherent has changed just the networking layer is websocket now.


This improved upon the existing http networking, now our golang process can send message to orchestrator (posing as websocket client) there is a simple interface for messages implemented.

Right now we dont send a lot of info from compiler using the websockets, only sending critical errors propagating in hook invocation and their success message.


Earlier we sent `taskId` and `pluginDir` as headers

```
{
  method: 'POST',
  headers: {
	  'Content-Type': 'application/json',
	  'X-Task-ID': task_id?.toString() ?? '',
	  'X-Plugin-Directory': directory,
  },
  body: JSON.stringify(payload),
},
```

```
type WebSocketMessage struct {
	ID              string         `json:"id"`
	Type            string         `json:"type"`
	Hook            string         `json:"hook"`
	Payload         map[string]any `json:"payload"`
	TaskID          string         `json:"taskId"`
	PluginDirectory string         `json:"pluginDirectory"`
}

```

Now we can just send them as regular payload.

We have a lot of errors already propagated to hook layer and then being sent to client, looks somewhat liks this 
<img width="2015" height="732" alt="image" src="https://github.com/user-attachments/assets/1c2121c9-8856-4624-9047-d5e812b575bb" />

siddarthvader committed:
websocket implementation

moved from http networking to websocket for hooks, bidirectional communication.

siddarthvader committed:
detaile fatal erorr and uuid for message id

siddarthvader committed:
rinsing same connection

siddarthvader committed:
remvoed deadling for read

siddarthvader committed:
delete request on timeout

siddarthvader committed:
kiling plugin process once the orchstatro is gone

siddarthvader committed:
sending paylaoid to all handler, removed duplication

siddarthvader committed:
added ws deps

siddarthvader committed:
using same error propagation as http handler

siddarthvader committed:
erorr fofrmatting fixed

siddarthvader committed:
Merge upstream go branch: Svelte plugin (#1556)

AlecAivazis commented on packages/houdini/src/lib/codegen.ts:
```diff
@@ -217,36 +220,125 @@ export async function codegen_setup(
+
+		return new Promise((resolve, reject) => {
+			// get a new connection
+			const wsUrl = `ws://localhost:${port}/ws`
```
Let's make the root of the plugin web server be the websocket endpoint

AlecAivazis commented on packages/houdini/src/lib/codegen.ts:
```diff
@@ -217,36 +220,125 @@ export async function codegen_setup(
+		const ws = await getOrCreateWS(name, port)
+
+		return new Promise((resolve, reject) => {
+			const messageId = `${hook}-${Date.now()}-${randomUUID()}`
```
Since there is only ever one orchestrator for a given plugin (one source of messages), I think we can simplify this a bit and define a `messageID` variable in global scope that we increment on every message.

AlecAivazis commented on packages/houdini/src/lib/codegen.ts:
```diff
@@ -217,36 +220,125 @@ export async function codegen_setup(
+
+			ws.on('close', () => {
+				// Remove from pool so next request creates new connection
+				// requestsd will eventually timeout and be rejected, we can agressively remove them but it's not a big deal
```
```suggestion
				// requests will eventually timeout and be rejected, we can agressively remove them but it's not a big deal
```

siddarthvader replied:
done

AlecAivazis commented on plugins/context.go:
```diff
@@ -28,6 +34,35 @@ func PluginDirFromContext(ctx context.Context) string {
+// 		Error: "TEST: This is a simulated non-fatal error during generation",
+// 	})
+// }
+// helper to get the conn from context,
```
```suggestion
// helper to get the conn from context
//
// Example:
//  	if conn := WSConnFromContext(ctx); conn != nil {
// 			conn.WriteJSON(WebSocketResponse{
// 				ID:    ctx.Value("wsMessageID").(string),
// 				Type:  "1error",
// 				Error: "TEST: This is a simulated non-fatal error during generation",
// 			})
// 		}
```

siddarthvader replied:
done

AlecAivazis commented on plugins/context.go:
```diff
@@ -28,6 +34,35 @@ func PluginDirFromContext(ctx context.Context) string {
 
-type taskIDCtxKey struct{}
+func ContextWithWSConn(ctx context.Context, conn *websocket.Conn) context.Context {
+	return context.WithValue(ctx, "wsConn", conn)
```
let's use a non-string key for the context here. Like you did for `taskIDCtxKey`

siddarthvader replied:
done

AlecAivazis commented on plugins/run.go:
```diff
@@ -55,12 +56,37 @@ func Run[PluginConfig any](plugin HoudiniPlugin[PluginConfig]) error {
+		defer conn.Close()
+
+		// connection established
+		log.Printf("ws connection established from %s", r.RemoteAddr)
```
This comment reminds me that we need to have a good story for different levels of logs. This kind of message is really helpful for debugging but something we probably don't want to show the user.

siddarthvader replied:
yes we could expose interface with log-levels, most def something we can implement. will leave a TODO comment.

AlecAivazis commented on plugins/websocket.go:
```diff
@@ -0,0 +1,471 @@
+	plugin HoudiniPlugin[PluginConfig],
+) func(ctx context.Context, payload map[string]any) (any, error) {
+	return func(ctx context.Context, payload map[string]any) (any, error) {
+		// if conn := WSConnFromContext(ctx); conn != nil {
```
let's remove this comment

siddarthvader replied:
done

AlecAivazis commented on plugins/websocket.go:
```diff
@@ -0,0 +1,471 @@
+
+	// --- AfterLoad is triggered for StaticRuntime OR AfterLoad
+	_, isStaticRuntime := plugin.(StaticRuntime)
+	_, isAfterLoad := plugin.(AfterLoad)
```
This file is missing some of the changes to the hook logic that was added as part of the svelte plugin PR. The one that stood out to me here was the `DefaultConfig` interface but there could be others

siddarthvader replied:
ported the remaining ones from http.go

AlecAivazis reviewed (changes requested):
Overall this looks really good! It's a significant upgrade from the current setup so thanks for pushing on this.

siddarthvader committed:
change websocker host and removed comments

siddarthvader committed:
poortred remaining hooks form http.go

AlecAivazis committed:
Merge branch 'go' of github.com:HoudiniGraphql/houdini into go

AlecAivazis committed:
fix build issue

siddarthvader committed:
addeds example for context helper

siddarthvader commented:
@AlecAivazis fixes pushed for all comments, kindly review.

AlecAivazis committed:
don't error if directory exists

AlecAivazis committed:
Merge branch 'go' of github.com:HoudiniGraphql/houdini into go

AlecAivazis commented on packages/houdini/src/lib/codegen.ts:
```diff
@@ -219,36 +221,127 @@ export async function codegen_setup(
-			body: JSON.stringify(payload),
-		})
+		// All hooks now use WebSocket
+		return await invoke_hook_websocket(name, hook, payload, task_id, port, directory)
```
let's remove the `_websocket` and stick with just `invoke_hook` since there is only a websocket version now

siddarthvader replied:
done

siddarthvader committed:
Merge branch 'HoudiniGraphql:go' into go

siddarthvader committed:
removed redundant wrapper function

siddarthvader committed:
using ws for triggerserial/parallel hook, formatting fixed

siddarthvader committed:
graacefully shiuting down connection in handleserial/parllel hoook

siddarthvader committed:
using file system plugin isntead of afero, ported from http.goq

siddarthvader committed:
diffenrece between plugin to plugin closure and orchestrator closure

using websocker status 1000a nd 1001 to differncitate, because in case
of orchstrator going away we need to kill the underlyuing process else
there is memory leak

siddarthvader committed:
made generate runtime un parallel_safe

siddarthvader committed:
checking procdss before killing and rmoved logs for websocket init

siddarthvader committed:
both httppa and websocker layers

abstracted hook hjandlers from websdocket and now both http and
websocket accept connection for all hooks implemented, also adding await
in cleanup of on_close in generate comamnd, it was leaking

siddarthvader committed:
prettier

siddarthvader committed:
registerting hooks in single pass

AlecAivazis reviewed (approved):
Thanks for getting this together! I think its finally good to go 🚀
